File: KCD/HMSuppliesView.m
//
//  HMSuppliesView.m
//  UITest
//
//  Created by Hori,Masaki on 2014/08/10.
//  Copyright (c) 2014年 Hori,Masaki. All rights reserved.
//

#import "HMSuppliesView.h"
#import "HMSuppliesCell.h"

#import "HMKCShipObject+Extensions.h"



static void *ShipStatusContext = &ShipStatusContext;


@interface HMSuppliesView ()
@property (nonatomic, strong) HMSuppliesCell *suppliesCell;
@end


@implementation HMSuppliesView


- (id)initWithFrame:(NSRect)frame
{
	self = [super initWithFrame:frame];
    if (self) {
		_suppliesCell = [HMSuppliesCell new];
		[self setCell:self.suppliesCell];
    }
    return self;
}

- (void)dealloc
{
	[self.suppliesCell.shipStatus removeObserver:self forKeyPath:@"fuel"];
	[self.suppliesCell.shipStatus removeObserver:self forKeyPath:@"maxFuel"];
	[self.suppliesCell.shipStatus removeObserver:self forKeyPath:@"bull"];
	[self.suppliesCell.shipStatus removeObserver:self forKeyPath:@"maxBull"];
}

- (void)setShipStatus:(HMKCShipObject *)shipStatus
{
	[self.suppliesCell.shipStatus removeObserver:self forKeyPath:@"fuel"];
	[self.suppliesCell.shipStatus removeObserver:self forKeyPath:@"maxFuel"];
	[self.suppliesCell.shipStatus removeObserver:self forKeyPath:@"bull"];
	[self.suppliesCell.shipStatus removeObserver:self forKeyPath:@"maxBull"];
	
	self.suppliesCell.shipStatus = shipStatus;
	
	[self.suppliesCell.shipStatus addObserver:self
								   forKeyPath:@"fuel"
									  options:NSKeyValueObservingOptionNew
									  context:ShipStatusContext];
	[self.suppliesCell.shipStatus addObserver:self
								   forKeyPath:@"maxFuel"
									  options:NSKeyValueObservingOptionNew
									  context:ShipStatusContext];
	[self.suppliesCell.shipStatus addObserver:self
								   forKeyPath:@"bull"
									  options:NSKeyValueObservingOptionNew
									  context:ShipStatusContext];
	[self.suppliesCell.shipStatus addObserver:self
								   forKeyPath:@"maxBull"
									  options:NSKeyValueObservingOptionNew
									  context:ShipStatusContext];
	
	[self setNeedsDisplay];
}
- (HMKCShipObject *)shipStatus
{
	return self.suppliesCell.shipStatus;
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
	if(context ==ShipStatusContext) {
		[self setNeedsDisplay];
		return;
	}
	
	[super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
}

@end

File: KCD/HMSuppliesView.swift
//
//  HMSuppliesView.swift
//  KCD
//
//  Created by Hori,Masaki on 2014/12/29.
//  Copyright (c) 2014年 Hori,Masaki. All rights reserved.
//

import Cocoa

class HMSuppliesView: NSControl
{
	override init(frame: NSRect) {
		suppliesCell = HMSuppliesCell()
		super.init(frame: frame)
		setCell(suppliesCell)
	}
	required init?(coder: NSCoder) {
		suppliesCell = HMSuppliesCell()
		super.init(coder: coder)
		setCell(suppliesCell)
	}
	
	let suppliesCell: HMSuppliesCell;
	
	var shipStatus: HMKCShipObject? {
		get {
			return suppliesCell.shipStatus
		}
		set {
			suppliesCell.shipStatus?.removeObserver(self, forKeyPath: "fuel")
			suppliesCell.shipStatus?.removeObserver(self, forKeyPath: "maxFuel")
			suppliesCell.shipStatus?.removeObserver(self, forKeyPath: "bull")
			suppliesCell.shipStatus?.removeObserver(self, forKeyPath: "maxBull")
			
			suppliesCell.shipStatus = newValue
			
			if let shipStatus = suppliesCell.shipStatus {
				shipStatus.addObserver(self, forKeyPath: "fuel", options: .New, context: nil)
				shipStatus.addObserver(self, forKeyPath: "maxFuel", options: .New, context: nil)
				shipStatus.addObserver(self, forKeyPath: "bull", options: .New, context: nil)
				shipStatus.addObserver(self, forKeyPath: "maxBull", options: .New, context: nil)
			}
			setNeedsDisplay()
		}
	}
	
	override func observeValueForKeyPath(keyPath: String, ofObject object: AnyObject, change: [NSObject : AnyObject], context: UnsafeMutablePointer<Void>) {
		switch keyPath {
		case "fuel", "maxFuel", "bull", "maxBull":
			setNeedsDisplay()
		default:
			super.observeValueForKeyPath(keyPath, ofObject: object, change: change, context: context)
		}
	}
}
